File: android/src/main/java/expo/modules/vungle/ReactNativeVungleModule.kt
package expo.modules.vungle

import android.util.Log
import com.vungle.ads.AdConfig
import com.vungle.ads.BaseAd
import com.vungle.ads.InitializationListener
import com.vungle.ads.RewardedAd
import com.vungle.ads.RewardedAdListener
import com.vungle.ads.VungleAds
import com.vungle.ads.VungleError
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

private const val TAG = "ReactNativeVungle"
private const val EVENT = "onVungle"

class ReactNativeVungleModule : Module() {
  @Volatile
  private var loggingEnabled: Boolean = true

  @Volatile
  private var initSucceeded: Boolean = false

  private val rewardedAds = ConcurrentHashMap<String, RewardedAd>()
  private val loadContinuations = ConcurrentHashMap<String, Continuation<Unit>>()
  private val showContinuations = ConcurrentHashMap<String, Continuation<Unit>>()

  private fun logNative(step: String, detail: String? = null) {
    if (!loggingEnabled) {
      return
    }
    val msg = if (detail == null) step else "$step — $detail"
    Log.i(TAG, msg)
  }

  private fun emit(kind: String, placementId: String? = null, extra: Map<String, Any?> = emptyMap()) {
    val payload = HashMap<String, Any?>(extra.size + 2)
    payload["kind"] = kind
    if (placementId != null) {
      payload["placementId"] = placementId
    }
    payload.putAll(extra)
    sendEvent(EVENT, payload)
  }

  private fun rewardedListener(placementId: String): RewardedAdListener {
    return object : RewardedAdListener {
      override fun onAdLoaded(baseAd: BaseAd) {
        logNative("rewarded_onAdLoaded", placementId)
        emit("rewardedAdLoaded", placementId)
        finishLoadContinuation(placementId, null)
      }

      override fun onAdStart(baseAd: BaseAd) {
        logNative("rewarded_onAdStart", placementId)
        emit("rewardedAdStarted", placementId)
      }

      override fun onAdImpression(baseAd: BaseAd) {
        logNative("rewarded_onAdImpression", placementId)
        emit("rewardedAdImpression", placementId)
      }

      override fun onAdEnd(baseAd: BaseAd) {
        logNative("rewarded_onAdEnd", placementId)
        emit("rewardedAdEnded", placementId)
        finishShowContinuation(placementId, null)
      }

      override fun onAdClicked(baseAd: BaseAd) {
        logNative("rewarded_onAdClicked", placementId)
        emit("rewardedAdClicked", placementId)
      }

      override fun onAdRewarded(baseAd: BaseAd) {
        logNative("rewarded_onAdRewarded", placementId)
        emit("rewardedUserRewarded", placementId)
      }

      override fun onAdLeftApplication(baseAd: BaseAd) {
        logNative("rewarded_onAdLeftApplication", placementId)
        emit("rewardedAdLeftApplication", placementId)
      }

      override fun onAdFailedToLoad(baseAd: BaseAd, adError: VungleError) {
        val code = adError.code.toString()
        val msg = adError.localizedMessage ?: "load_failed"
        logNative("rewarded_onAdFailedToLoad", "$placementId $code $msg")
        emit(
          "rewardedAdFailedToLoad",
          placementId,
          mapOf("code" to code, "message" to msg)
        )
        finishLoadContinuation(
          placementId,
          CodedException("ERR_VUNGLE_LOAD_FAILED", "$code: $msg", null)
        )
      }

      override fun onAdFailedToPlay(baseAd: BaseAd, adError: VungleError) {
        val code = adError.code.toString()
        val msg = adError.localizedMessage ?: "play_failed"
        logNative("rewarded_onAdFailedToPlay", "$placementId $code $msg")
        emit(
          "rewardedAdFailedToPlay",
          placementId,
          mapOf("code" to code, "message" to msg)
        )
        finishShowContinuation(
          placementId,
          CodedException("ERR_VUNGLE_PLAY_FAILED", "$code: $msg", null)
        )
      }
    }
  }

  private fun finishLoadContinuation(placementId: String, error: Throwable?) {
    val cont = loadContinuations.remove(placementId) ?: return
    if (error == null) {
      try {
        cont.resume(Unit)
      } catch (_: Throwable) {
      }
    } else {
      try {
        cont.resumeWithException(error)
      } catch (_: Throwable) {
      }
    }
  }

  private fun finishShowContinuation(placementId: String, error: Throwable?) {
    val cont = showContinuations.remove(placementId) ?: return
    try {
      if (error == null) {
        cont.resume(Unit)
      } else {
        cont.resumeWithException(error)
      }
    } catch (_: Throwable) {
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ReactNativeVungle")

    Events(EVENT)

    Function("getSdkVersionSync") {
      try {
        VungleAds.getSdkVersion()
      } catch (e: Throwable) {
        this@ReactNativeVungleModule.logNative("getSdkVersionSync_error", e.message)
        "unknown"
      }
    }

    Function("setLoggingEnabled") { enabled: Boolean ->
      this@ReactNativeVungleModule.loggingEnabled = enabled
      this@ReactNativeVungleModule.logNative("setLoggingEnabled", enabled.toString())
    }

    AsyncFunction("initSdkAsync") Coroutine { appId: String, loggingEnabledArg: Boolean ->
      val host = this@ReactNativeVungleModule
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      host.loggingEnabled = loggingEnabledArg
      host.logNative("initSdkAsync_start", "appIdLength=${appId.length}")

      if (host.initSucceeded) {
        host.logNative("initSdkAsync_skip", "already_initialized")
        host.emit("sdkInitSkipped", extra = mapOf("reason" to "already_initialized"))
        return@Coroutine true
      }

      suspendCoroutine { cont ->
        VungleAds.init(context, appId, object : InitializationListener {
          override fun onSuccess() {
            host.initSucceeded = true
            host.logNative("initSdkAsync_success")
            host.emit("sdkInitialized", extra = mapOf("success" to true))
            cont.resume(Unit)
          }

          override fun onError(vungleError: VungleError) {
            val code = vungleError.code.toString()
            val msg = vungleError.localizedMessage ?: "init_error"
            host.logNative("initSdkAsync_error", "$code $msg")
            host.emit(
              "sdkInitialized",
              extra = mapOf(
                "success" to false,
                "code" to code,
                "message" to msg
              )
            )
            cont.resumeWithException(
              CodedException("ERR_VUNGLE_INIT", "$code: $msg", null)
            )
          }
        })
      }
      true
    }

    AsyncFunction("loadRewardedAsync") Coroutine { placementId: String, userId: String? ->
      val host = this@ReactNativeVungleModule
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      if (!host.initSucceeded) {
        throw CodedException("ERR_VUNGLE_NOT_INITIALIZED", "Call initSdkAsync first", null)
      }

      host.rewardedAds.remove(placementId)?.adListener = null

      host.logNative("loadRewarded_start", placementId)
      host.emit("rewardedLoadStarted", placementId)

      suspendCoroutine { cont ->
        host.loadContinuations[placementId]?.let { old ->
          host.loadContinuations.remove(placementId)
          old.resumeWithException(CodedException("ERR_VUNGLE_LOAD_REPLACED", "Superseded by new load", null))
        }
        host.loadContinuations[placementId] = cont

        val ad = RewardedAd(context, placementId, AdConfig()).apply {
          if (!userId.isNullOrBlank()) {
            setUserId(userId)
          }
          adListener = host.rewardedListener(placementId)
          load(null)
        }
        host.rewardedAds[placementId] = ad
      }
      true
    }

    AsyncFunction("showRewardedAsync") Coroutine { placementId: String ->
      val host = this@ReactNativeVungleModule
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      if (!host.initSucceeded) {
        throw CodedException("ERR_VUNGLE_NOT_INITIALIZED", "Call initSdkAsync first", null)
      }
      val ad = host.rewardedAds[placementId]
        ?: throw CodedException("ERR_VUNGLE_AD_MISSING", "No ad for placement $placementId", null)

      host.logNative("showRewarded_check", placementId)
      if (ad.canPlayAd() != true) {
        host.emit(
          "rewardedAdShowRejected",
          placementId,
          mapOf("reason" to "canPlayAd_false")
        )
        throw CodedException("ERR_VUNGLE_CANNOT_PLAY", "canPlayAd() is false for $placementId", null)
      }

      host.emit("rewardedAdWillShow", placementId)

      suspendCoroutine { cont ->
        host.showContinuations[placementId]?.let { old ->
          host.showContinuations.remove(placementId)
          old.resumeWithException(CodedException("ERR_VUNGLE_SHOW_REPLACED", "Superseded by new show", null))
        }
        host.showContinuations[placementId] = cont
        ad.play(context)
      }
      true
    }

    AsyncFunction("destroyRewardedAsync") Coroutine { placementId: String ->
      val host = this@ReactNativeVungleModule
      host.rewardedAds.remove(placementId)?.let { ad ->
        ad.adListener = null
        host.logNative("destroyRewarded", placementId)
        host.emit("rewardedAdDestroyed", placementId)
      }
      host.finishLoadContinuation(
        placementId,
        CodedException("ERR_VUNGLE_DESTROYED", "Ad destroyed", null)
      )
      host.finishShowContinuation(
        placementId,
        CodedException("ERR_VUNGLE_DESTROYED", "Ad destroyed", null)
      )
      true
    }
  }
}

File: ios/ReactNativeVungleModule.swift
import ExpoModulesCore
import Foundation
import UIKit
import VungleAdsSDK

private let vungleEventName = "onVungle"
private let logTag = "ReactNativeVungle"

private func topViewController() -> UIViewController? {
  let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
  for scene in scenes where scene.activationState == .foregroundActive {
    if let root = scene.windows.first(where: { $0.isKeyWindow })?.rootViewController {
      var top = root
      while let presented = top.presentedViewController {
        top = presented
      }
      return top
    }
  }
  if let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
     let root = scene.windows.first?.rootViewController {
    var top = root
    while let presented = top.presentedViewController {
      top = presented
    }
    return top
  }
  return nil
}

private final class RewardedBridge: NSObject, VungleRewardedDelegate {
  weak var owner: ReactNativeVungleModule?
  let placementId: String

  init(owner: ReactNativeVungleModule, placementId: String) {
    self.owner = owner
    self.placementId = placementId
  }

  func rewardedAdDidLoad(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdDidLoad", detail: placementId)
    owner?.emit(kind: "rewardedAdLoaded", placementId: placementId)
    owner?.finishLoadContinuation(placementId: placementId, error: nil)
  }

  func rewardedAdDidFailToLoad(_ rewarded: VungleRewarded, withError: NSError) {
    owner?.logNative("rewardedAdDidFailToLoad", detail: "\(placementId) \(withError.code) \(withError.localizedDescription)")
    owner?.emit(
      kind: "rewardedAdFailedToLoad",
      placementId: placementId,
      extra: ["code": String(withError.code), "message": withError.localizedDescription]
    )
    owner?.finishLoadContinuation(
      placementId: placementId,
      error: Exception(
        name: "VungleLoadFailed",
        description: withError.localizedDescription,
        code: "ERR_VUNGLE_LOAD_FAILED"
      )
    )
  }

  func rewardedAdWillPresent(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdWillPresent", detail: placementId)
    owner?.emit(kind: "rewardedAdWillPresent", placementId: placementId)
  }

  func rewardedAdDidPresent(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdDidPresent", detail: placementId)
    owner?.emit(kind: "rewardedAdStarted", placementId: placementId)
  }

  func rewardedAdDidFailToPresent(_ rewarded: VungleRewarded, withError: NSError) {
    owner?.logNative("rewardedAdDidFailToPresent", detail: "\(placementId) \(withError.code)")
    owner?.emit(
      kind: "rewardedAdFailedToPlay",
      placementId: placementId,
      extra: ["code": String(withError.code), "message": withError.localizedDescription]
    )
    owner?.finishShowContinuation(
      placementId: placementId,
      error: Exception(
        name: "VunglePlayFailed",
        description: withError.localizedDescription,
        code: "ERR_VUNGLE_PLAY_FAILED"
      )
    )
  }

  func rewardedAdDidTrackImpression(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdDidTrackImpression", detail: placementId)
    owner?.emit(kind: "rewardedAdImpression", placementId: placementId)
  }

  func rewardedAdDidClick(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdDidClick", detail: placementId)
    owner?.emit(kind: "rewardedAdClicked", placementId: placementId)
  }

  func rewardedAdWillLeaveApplication(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdWillLeaveApplication", detail: placementId)
    owner?.emit(kind: "rewardedAdLeftApplication", placementId: placementId)
  }

  func rewardedAdDidRewardUser(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdDidRewardUser", detail: placementId)
    owner?.emit(kind: "rewardedUserRewarded", placementId: placementId)
  }

  func rewardedAdWillClose(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdWillClose", detail: placementId)
    owner?.emit(kind: "rewardedAdWillClose", placementId: placementId)
  }

  func rewardedAdDidClose(_ rewarded: VungleRewarded) {
    owner?.logNative("rewardedAdDidClose", detail: placementId)
    owner?.emit(kind: "rewardedAdEnded", placementId: placementId)
    owner?.finishShowContinuation(placementId: placementId, error: nil)
  }
}

public final class ReactNativeVungleModule: Module {
  private let stateLock = NSLock()
  private var loggingEnabled = true
  private var initSucceeded = false
  private var rewardedAds: [String: VungleRewarded] = [:]
  private var bridges: [String: RewardedBridge] = [:]
  private var loadContinuations: [String: CheckedContinuation<Void, Error>] = [:]
  private var showContinuations: [String: CheckedContinuation<Void, Error>] = [:]

  fileprivate func logNative(_ step: String, detail: String? = nil) {
    stateLock.lock()
    let enabled = loggingEnabled
    stateLock.unlock()
    if !enabled {
      return
    }
    if let detail = detail {
      NSLog("[%@] %@ — %@", logTag, step, detail)
    } else {
      NSLog("[%@] %@", logTag, step)
    }
  }

  fileprivate func emit(kind: String, placementId: String? = nil, extra: [String: Any] = [:]) {
    var body: [String: Any?] = ["kind": kind]
    if let placementId = placementId {
      body["placementId"] = placementId
    }
    for (k, v) in extra {
      body[k] = v
    }
    sendEvent(vungleEventName, body)
  }

  fileprivate func finishLoadContinuation(placementId: String, error: Error?) {
    stateLock.lock()
    let cont = loadContinuations.removeValue(forKey: placementId)
    stateLock.unlock()
    guard let cont = cont else {
      return
    }
    if let error = error {
      cont.resume(throwing: error)
    } else {
      cont.resume()
    }
  }

  fileprivate func finishShowContinuation(placementId: String, error: Error?) {
    stateLock.lock()
    let cont = showContinuations.removeValue(forKey: placementId)
    stateLock.unlock()
    guard let cont = cont else {
      return
    }
    if let error = error {
      cont.resume(throwing: error)
    } else {
      cont.resume()
    }
  }

  public func definition() -> ModuleDefinition {
    Name("ReactNativeVungle")

    Events(vungleEventName)

    Function("getSdkVersionSync") { () -> String in
      VungleAds.sdkVersion
    }

    Function("setLoggingEnabled") { (enabled: Bool) in
      self.stateLock.lock()
      self.loggingEnabled = enabled
      self.stateLock.unlock()
      self.logNative("setLoggingEnabled", detail: String(enabled))
    }

    AsyncFunction("initSdkAsync") { (appId: String, loggingEnabledArg: Bool) async throws -> Bool in
      self.stateLock.lock()
      self.loggingEnabled = loggingEnabledArg
      self.stateLock.unlock()
      self.logNative("initSdkAsync_start", detail: "appIdLength=\(appId.count)")

      if VungleAds.isInitialized() {
        self.stateLock.lock()
        self.initSucceeded = true
        self.stateLock.unlock()
        self.logNative("initSdkAsync_skip", detail: "already_initialized")
        self.emit(kind: "sdkInitSkipped", extra: ["reason": "already_initialized"])
        return true
      }

      return try await withCheckedThrowingContinuation { continuation in
        VungleAds.initWithAppId(appId) { error in
          if let error = error {
            let nsError = error as NSError
            self.logNative("initSdkAsync_error", detail: "\(nsError.code) \(error.localizedDescription)")
            self.emit(
              kind: "sdkInitialized",
              extra: [
                "success": false,
                "code": String(nsError.code),
                "message": error.localizedDescription
              ]
            )
            continuation.resume(
              throwing: Exception(
                name: "VungleInitFailed",
                description: error.localizedDescription,
                code: "ERR_VUNGLE_INIT"
              )
            )
          } else {
            self.stateLock.lock()
            self.initSucceeded = true
            self.stateLock.unlock()
            self.logNative("initSdkAsync_success")
            self.emit(kind: "sdkInitialized", extra: ["success": true])
            continuation.resume(returning: true)
          }
        }
      }
    }

    AsyncFunction("loadRewardedAsync") { (placementId: String, userId: String?) async throws -> Bool in
      self.stateLock.lock()
      let initialized = self.initSucceeded
      self.stateLock.unlock()
      if !initialized {
        throw Exception(
          name: "VungleNotInitialized",
          description: "Call initSdkAsync first",
          code: "ERR_VUNGLE_NOT_INITIALIZED"
        )
      }

      self.stateLock.lock()
      if let old = self.rewardedAds.removeValue(forKey: placementId) {
        old.delegate = nil
      }
      self.bridges.removeValue(forKey: placementId)
      self.stateLock.unlock()

      self.logNative("loadRewarded_start", detail: placementId)
      self.emit(kind: "rewardedLoadStarted", placementId: placementId)

      try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        self.stateLock.lock()
        if let oldCont = self.loadContinuations.removeValue(forKey: placementId) {
          self.stateLock.unlock()
          oldCont.resume(
            throwing: Exception(
              name: "VungleLoadReplaced",
              description: "Superseded by new load",
              code: "ERR_VUNGLE_LOAD_REPLACED"
            )
          )
          self.stateLock.lock()
        }
        self.loadContinuations[placementId] = continuation
        self.stateLock.unlock()

        let rewarded = VungleRewarded(placementId: placementId)
        let bridge = RewardedBridge(owner: self, placementId: placementId)
        rewarded.delegate = bridge
        if let userId = userId, !userId.isEmpty {
          rewarded.setUserIdWithUserId(userId)
          self.logNative("loadRewarded_userId", detail: "setUserIdWithUserId len=\(userId.count)")
        }

        self.stateLock.lock()
        self.rewardedAds[placementId] = rewarded
        self.bridges[placementId] = bridge
        self.stateLock.unlock()

        rewarded.load()
      }

      return true
    }

    AsyncFunction("showRewardedAsync") { (placementId: String) async throws -> Bool in
      self.stateLock.lock()
      let initialized = self.initSucceeded
      let rewarded = self.rewardedAds[placementId]
      self.stateLock.unlock()

      if !initialized {
        throw Exception(
          name: "VungleNotInitialized",
          description: "Call initSdkAsync first",
          code: "ERR_VUNGLE_NOT_INITIALIZED"
        )
      }
      guard let rewarded = rewarded else {
        throw Exception(
          name: "VungleAdMissing",
          description: "No ad for placement \(placementId)",
          code: "ERR_VUNGLE_AD_MISSING"
        )
      }

      self.logNative("showRewarded_check", detail: placementId)
      guard rewarded.canPlayAd() else {
        self.emit(
          kind: "rewardedAdShowRejected",
          placementId: placementId,
          extra: ["reason": "canPlayAd_false"]
        )
        throw Exception(
          name: "VungleCannotPlay",
          description: "canPlayAd() is false for \(placementId)",
          code: "ERR_VUNGLE_CANNOT_PLAY"
        )
      }

      self.emit(kind: "rewardedAdWillShow", placementId: placementId)

      guard let presenter = topViewController() else {
        throw Exception(
          name: "VungleNoViewController",
          description: "Could not find a UIViewController to present the ad",
          code: "ERR_VUNGLE_NO_VIEW_CONTROLLER"
        )
      }

      try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        self.stateLock.lock()
        if let oldCont = self.showContinuations.removeValue(forKey: placementId) {
          self.stateLock.unlock()
          oldCont.resume(
            throwing: Exception(
              name: "VungleShowReplaced",
              description: "Superseded by new show",
              code: "ERR_VUNGLE_SHOW_REPLACED"
            )
          )
          self.stateLock.lock()
        }
        self.showContinuations[placementId] = continuation
        self.stateLock.unlock()

        rewarded.present(with: presenter)
      }

      return true
    }

    AsyncFunction("destroyRewardedAsync") { (placementId: String) async throws -> Bool in
      self.stateLock.lock()
      if let rewarded = self.rewardedAds.removeValue(forKey: placementId) {
        rewarded.delegate = nil
        self.bridges.removeValue(forKey: placementId)
        self.stateLock.unlock()
        self.logNative("destroyRewarded", detail: placementId)
        self.emit(kind: "rewardedAdDestroyed", placementId: placementId)
      } else {
        self.stateLock.unlock()
      }

      self.finishLoadContinuation(
        placementId: placementId,
        error: Exception(
          name: "VungleDestroyed",
          description: "Ad destroyed",
          code: "ERR_VUNGLE_DESTROYED"
        )
      )
      self.finishShowContinuation(
        placementId: placementId,
        error: Exception(
          name: "VungleDestroyed",
          description: "Ad destroyed",
          code: "ERR_VUNGLE_DESTROYED"
        )
      )
      return true
    }
  }
}
